File: src/main/java/com/polygloat/controllers/UserAppApiController.java
package com.polygloat.controllers;

import com.polygloat.constants.ApiScope;
import com.polygloat.constants.Message;
import com.polygloat.dtos.PathDTO;
import com.polygloat.dtos.request.SetTranslationsDTO;
import com.polygloat.exceptions.NotFoundException;
import com.polygloat.model.ApiKey;
import com.polygloat.model.Language;
import com.polygloat.model.Repository;
import com.polygloat.model.Source;
import com.polygloat.security.AuthenticationFacade;
import com.polygloat.service.*;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.*;
import java.util.stream.Collectors;

@RestController
@CrossOrigin(origins = "*")
@RequestMapping("/uaa")
@RequiredArgsConstructor(onConstructor = @__(@Autowired))
public class UserAppApiController implements IController {

    private final TranslationService translationService;
    private final SourceService sourceService;
    private final RepositoryService repositoryService;
    private final SecurityService securityService;
    private final AuthenticationFacade authenticationFacade;
    private final LanguageService languageService;

    @GetMapping(value = "/{languages}")
    public Map<String, Object> getTranslations(@PathVariable("languages") String languages) {
        ApiKey apiKey = authenticationFacade.getApiKey();
        securityService.checkApiKeyScopes(Set.of(ApiScope.TRANSLATIONS_VIEW), apiKey);
        return translationService.getTranslations(parseLanguages(languages).orElse(null), apiKey.getRepository().getId());
    }

    @GetMapping(value = "/source/{sourceFullPath}/{languages}")
    public Map<String, String> getSourceTranslations(@PathVariable("sourceFullPath") String fullPath,
                                                     @PathVariable("languages") String langs) {
        PathDTO pathDTO = PathDTO.fromFullPath(fullPath);
        ApiKey apiKey = authenticationFacade.getApiKey();
        securityService.checkApiKeyScopes(Set.of(ApiScope.TRANSLATIONS_VIEW), apiKey);
        return translationService.getSourceTranslationsResult(apiKey.getRepository().getId(), pathDTO, parseLanguages(langs).orElse(null));
    }

    @GetMapping(value = "/source/{sourceFullPath}")
    public Map<String, String> getSourceTranslations(@PathVariable("sourceFullPath") String fullPath) {
        PathDTO pathDTO = PathDTO.fromFullPath(fullPath);
        ApiKey apiKey = authenticationFacade.getApiKey();
        return translationService.getSourceTranslationsResult(apiKey.getRepository().getId(), pathDTO, null);
    }

    @PostMapping("")
    public void setTranslations(@RequestBody @Valid SetTranslationsDTO dto) {
        ApiKey apiKey = authenticationFacade.getApiKey();
        securityService.checkApiKeyScopes(Set.of(ApiScope.TRANSLATIONS_EDIT), apiKey);
        Repository repository = repositoryService.findById(apiKey.getRepository().getId()).orElseThrow(() -> new NotFoundException(Message.REPOSITORY_NOT_FOUND));
        Source source = sourceService.getOrCreateSource(repository, PathDTO.fromFullPath(dto.getSourceFullPath()));
        translationService.setForSource(source, dto.getTranslations());
    }

    @GetMapping("/languages")
    public Set<String> getLanguages() {
        ApiKey apiKey = authenticationFacade.getApiKey();
        securityService.checkApiKeyScopes(Set.of(ApiScope.TRANSLATIONS_EDIT), apiKey);
        return languageService.findAll(apiKey.getRepository().getId()).stream().map(Language::getAbbreviation).collect(Collectors.toSet());
    }

    @GetMapping("/scopes")
    public Set<String> getScopes() {
        ApiKey apiKey = authenticationFacade.getApiKey();
        return apiKey.getScopes().stream().map(ApiScope::getValue).collect(Collectors.toSet());
    }

    private Optional<Set<String>> parseLanguages(String languages) {
        if (languages == null) {
            return Optional.empty();
        }
        return Optional.of(new HashSet<>(Arrays.stream(
                languages.split(","))
                //filter out empty strings
                .filter(i ->
                        !i.isEmpty()
                )
                .collect(Collectors.toList())));
    }
}

File: src/main/kotlin/io/tolgee/controllers/UserAppApiController.kt
package io.tolgee.controllers

import io.swagger.v3.oas.annotations.Hidden
import io.tolgee.constants.ApiScope
import io.tolgee.constants.Message
import io.tolgee.dtos.PathDTO
import io.tolgee.dtos.request.GetKeyTranslationsReqDto
import io.tolgee.dtos.request.SetTranslationsDTO
import io.tolgee.exceptions.NotFoundException
import io.tolgee.model.Language
import io.tolgee.security.AuthenticationFacade
import io.tolgee.security.api_key_auth.AccessWithApiKey
import io.tolgee.service.*
import org.springframework.web.bind.annotation.*
import java.util.stream.Collectors
import javax.validation.Valid

@RestController
@CrossOrigin(origins = ["*"])
@RequestMapping("/uaa")
@Deprecated(message = "This endpoint is deprecated. Integration libraries should use standard endpoints for each resource.")
@Hidden
class UserAppApiController(
        private val translationService: TranslationService,
        private val keyService: KeyService,
        private val projectService: ProjectService,
        private val securityService: SecurityService,
        private val authenticationFacade: AuthenticationFacade,
        private val languageService: LanguageService,
) {
    @GetMapping(value = ["/{languages}"])
    @AccessWithApiKey
    @Deprecated(message = "Use standard /api/project/translations/...")
    fun getTranslations(@PathVariable("languages") languages: Set<String>): Map<String, Any> {
        val apiKey = authenticationFacade.apiKey
        securityService.checkApiKeyScopes(setOf(ApiScope.TRANSLATIONS_VIEW), apiKey)
        return translationService.getTranslations(languages, apiKey.project!!.id)
    }

    @GetMapping(value = ["/source/{key:.+}/{languages}", "/key/{key:.+}/{languages}"])
    @AccessWithApiKey
    @Deprecated("can not pass . as parameter of text, for longer texts it would be much better to use POST")
    fun getKeyTranslations(@PathVariable("key") fullPath: String?,
                           @PathVariable("languages") langs: Set<String>?): Map<String, String?> {
        val pathDTO = PathDTO.fromFullPath(fullPath)
        val apiKey = authenticationFacade.apiKey
        securityService.checkApiKeyScopes(setOf(ApiScope.TRANSLATIONS_VIEW), apiKey)
        return translationService.getKeyTranslationsResult(apiKey.project!!.id, pathDTO, langs)
    }

    @PostMapping(value = ["/keyTranslations/{languages}"])
    @AccessWithApiKey([ApiScope.TRANSLATIONS_VIEW])
    fun getKeyTranslationsPost(@RequestBody body: GetKeyTranslationsReqDto, @PathVariable("languages") langs: Set<String>?): Map<String, String?> {
        val pathDTO = PathDTO.fromFullPath(body.key)
        val apiKey = authenticationFacade.apiKey
        return translationService.getKeyTranslationsResult(apiKey.project!!.id, pathDTO, langs)
    }

    @GetMapping(value = ["/source/{key:.+}"])
    @AccessWithApiKey([ApiScope.TRANSLATIONS_VIEW])
    @Deprecated("can not pass . as parameter of text, for longer texts it would be much better to use POST")
    fun getKeyTranslations(@PathVariable("key") fullPath: String?): Map<String, String?> {
        val pathDTO = PathDTO.fromFullPath(fullPath)
        val apiKey = authenticationFacade.apiKey
        return translationService.getKeyTranslationsResult(apiKey.project!!.id, pathDTO, null)
    }

    @PostMapping(value = ["/keyTranslations"])
    @AccessWithApiKey([ApiScope.TRANSLATIONS_VIEW])
    fun getKeyTranslationsPost(@RequestBody body: GetKeyTranslationsReqDto): Map<String, String?> {
        val pathDTO = PathDTO.fromFullPath(body.key)
        val apiKey = authenticationFacade.apiKey
        return translationService.getKeyTranslationsResult(apiKey.project!!.id, pathDTO, null)
    }

    @PostMapping("")
    @AccessWithApiKey([ApiScope.TRANSLATIONS_EDIT])
    fun setTranslations(@RequestBody @Valid dto: SetTranslationsDTO) {
        val apiKey = authenticationFacade.apiKey
        val project = projectService.get(apiKey.project!!.id).orElseThrow { NotFoundException(Message.PROJECT_NOT_FOUND) }!!
        val key = keyService.getOrCreateKey(project, PathDTO.fromFullPath(dto.key))
        translationService.setForKey(key, dto.translations!!)
    }

    @Deprecated(message = "Use /api/project/languages")
    @AccessWithApiKey
    @GetMapping("/languages")
    fun getLanguages(): Set<String> {
        val apiKey = authenticationFacade.apiKey
        securityService.checkApiKeyScopes(setOf(ApiScope.TRANSLATIONS_EDIT), apiKey)
        return languageService.findAll(apiKey.project!!.id).stream().map { obj: Language -> obj.abbreviation!! }.collect(Collectors.toSet())
    }

    @AccessWithApiKey
    @GetMapping("/scopes")
    fun getScopes(): Set<String> {
        val apiKey = authenticationFacade.apiKey
        return apiKey.scopesEnum.asSequence().map { it.value }.toSet()
    }
}
